### ISSUE TITLE
Timeline group ordering performance suggestion 

### ISSUE THREAD

**Issue by ChenMachluf** [almende/vis#1207]
Hi,
I use vis timeline with 40k~ object that i update dynamically.
The performance bottleneck in vis is that is reorder all the group on every add/remove/update to the dataset.

I create a version of vis that instead of call group.order after every add/remove/update , 
it make sure that always the items is sorted.
i use binaryIndex to search the correct index to push the item in or to remove it.

i create 2 examples (based on visjs.org example) of timeline with 100000 items , with update and add function(1,1000,10000,10000 jumps)

original version : http://embed.plnkr.co/dioxgHCCj2TStrypDi9f/preview
my version : http://embed.plnkr.co/ujsM82uH2K3oannUCIYx/preview

As you can see to update one item in original version take more then a second.
In my version its take around milisecond.
But if we go to update 10000 or 100000 the original version is more faster , 
actually to the original version its almost the same time no depends of the items count (cuz it sort one time  all the items). 

Same is applied also to the add ,but just in the 100000 zone original version is faster.

this version solved my perf problems , cuz i update 400< items every few seconds.
my version make it slower if the update/add/remove is with big numbers.

I will love to pull request it if u think its better that way

**Comment by josdejong** [comment]
Hi Chen, that sounds like a nice improvement. A pull request would be great!

**Comment by josdejong** [comment]
btw I ran both your examples, and I don't get the expected performance improvements:

Chrome:
original version outputs "creation took 1553 mils"
improved version outputs "creation took 1295 mils"

Firefox:
original version outputs "creation took 1104 mils"
improved version outputs "creation took 1530 mils"

any idea?

**Comment by ChenMachluf** [comment]
wow , ff do Array.sort much faster then chrome.
and chrome is faster using splice.
I check it only on chrome.... /:
In ff the creation is more faster and if you update/insert more then 500 items it better use the existing version.

I think (no ideal) maybe use some of hybrid depends on the count of the change element (and maybe platform? or overkill?)

what you think?

**Comment by josdejong** [comment]
Thanks for your PR in #1211, that clarifies your solution. It's still not clear whether and when this actually yields an improved performance, I can't reproduce it.

Some sort of hybrid solution sounds indeed like overkill, unless it solves real performance issues for multiple real use cases. But if not, I prefer to keep the code as simple as possible.

If I understand your use case correctly, you do a lot of `dataset.update(changedItem)` calls. Can't you instead do a single call to the dataset, like `dataset.update(allChangedItems)`?

**Comment by ChenMachluf** [comment]
no , i cant.
i do a lot of dataset.update(allChangedItems).
but this happens like every 4-3 seconds , and it takes 1~ sec .

**Comment by josdejong** [comment]
Ok so the main bottleneck you try to tackle is the duration of `order`, which takes 2/3 of the time. Would be great to improve this.

I tried your PR, but I still didn't succeed in seeing performance improvements (your PR looks even slower but I may do something wrong). I did the following:

- Open `examples/timeline/other/groupsPerformance` in my browser
- Force an update of all items and measure duration in the browser console:

  ```js
  console.time('update'); items.update(items.get()); console.timeEnd('update')
  ```
- compare the duration with original code vs. your PR
